File: clients/go/pandora.go
package pandora

import (
	"encoding/json"

	zmq "github.com/alecthomas/gozmq"
)

var Address = "tcp://127.0.0.1:9999"
var socket *zmq.Socket

const (
	FLAG_HTML     = "html"
	FLAG_MARKDOWN = "markdown"
)

type Input struct {
	Text string `json:"text"`
	From string `json:"from"`
	To   string `json:"to"`
}

func init() {
	context, _ := zmq.NewContext()
	socket, _ = context.NewSocket(zmq.REQ)
	err := socket.Connect(Address)
	if err != nil {
		panic("Please make sure the pandora daemon is running")
	}
}

func ToMarkdown(html string) (r string, err error) {
	data := &Input{
		Text: html,
		From: FLAG_HTML,
		To:   FLAG_MARKDOWN,
	}

	return send(data)
}

func ToHTML(md string) (r string, err error) {
	data := &Input{
		Text: md,
		From: FLAG_MARKDOWN,
		To:   FLAG_HTML,
	}

	return send(data)
}

func send(data interface{}) (r string, err error) {
	var sendBytes []byte
	sendBytes, err = json.Marshal(data)
	if err != nil {
		return
	}

	err = socket.Send(sendBytes, 0)
	if err != nil {
		return
	}

	var b []byte
	b, err = socket.Recv(0)
	if err != nil {
		return
	}
	r = string(b)
	return
}

File: clients/ruby/pandora.rb
require 'rubygems'
require 'ffi-rzmq'
require 'json'

module Pandora
  def to_markdown text
    Converter.new.to_markdown text
  end

  def to_html text
    Converter.new.to_html text
  end

  class Converter
    FLAG_HTML = "html"
    FLAG_MARKDOWN = "markdown"
    def initialize
      context = ZMQ::Context.new(1)
      @requester = context.socket(ZMQ::REQ)
      @requester.connect("tcp://127.0.0.1:9999")
    end

    def to_markdown text 
      convert text, FLAG_HTML, FLAG_MARKDOWN
    end

    def to_html text
      convert text, FLAG_MARKDOWN, FLAG_HTML
    end

    private
    def convert(text, from, to)
      obj = {text: text, from: from, to: to}
      msg = obj.to_json
      @requester.send_string msg
      reply = ''
      @requester.recv_string(reply)
      reply
    end

  end
end
